devtools/server/actors/replay/pure-eval-dbg.jsm

```diff
@@ -0,0 +1,189 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+// The functions in this file have primarily been copied from
+// 'devtools/server/actors/webconsole/eval-with-debugger.js'. We had to
+// duplicate them in order to ensure that they run in the correct compartment.
+
+const { require } = ChromeUtils.import("resource://devtools/shared/Loader.jsm");
+
+const eagerEcmaAllowlist = require("devtools/server/actors/webconsole/eager-ecma-allowlist");
+const eagerFunctionAllowlist = require("devtools/server/actors/webconsole/eager-function-allowlist");
+
+const EXPORTED_SYMBOLS = ["evalExpression"];
+
+function evalExpression(dbg, dbgGlobal, frame, expression, pure) {
+  let noSideEffectDebugger = null;
+  if (pure) {
+    noSideEffectDebugger = makeSideeffectFreeDebugger(dbg);
+
+    // Bug 1637883 demonstrated an issue where dbgGlobal was somehow in the
+    // same compartment as the Debugger, meaning it could not be debugged
+    // and thus cannot handle eager evaluation. In that case we skip execution.
+    if (!noSideEffectDebugger.hasDebuggee(dbgGlobal.unsafeDereference())) {
+      return null;
+    }
+
+    // When a sideeffect-free debugger has been created, we need to eval
+    // in the context of that debugger in order for the side-effect tracking
+    // to apply.
+    frame = frame ? noSideEffectDebugger.adoptFrame(frame) : null;
+    dbgGlobal = noSideEffectDebugger.adoptDebuggeeValue(dbgGlobal);
+  }
+
+  let completion;
+  try {
+    if (frame) {
+      completion = frame.eval(expression);
+    } else {
+      completion = dbgGlobal.executeInGlobal(expression);
+    }
+
+    if (noSideEffectDebugger && completion) {
+      if ("return" in completion) {
+        completion.return = dbg.adoptDebuggeeValue(completion.return);
+      }
+      if ("throw" in completion) {
+        completion.throw = dbg.adoptDebuggeeValue(completion.throw);
+      }
+    }
+  } finally {
+    // We need to be absolutely sure that the sideeffect-free debugger's
+    // debuggees are removed because otherwise we risk them terminating
+    // execution of later code in the case of unexpected exceptions.
+    if (noSideEffectDebugger) {
+      noSideEffectDebugger.removeAllDebuggees();
+    }
+  }
+
+  return completion;
+}
+
+function makeSideeffectFreeDebugger(originalDbg) {
```

Comment:
Here we use `Sideeffect` but on line 72 we use `SideEffect`. I think I prefer `SideEffect` personally, but I think either is fine as long as it's consistent

Reply:
I'll just throw one comment here to address all three of these. Since this file is 90% copied from `devtools/server/actors/webconsole/eval-with-debugger.js`, all of these comments and the logic are existing code, so I don't think making changes to them is a good idea.

Reply:
Yeah I saw that patch when I was looking at the bug you mentioned, I agree 👍